src/ert/gui/model/node.py

```diff
@@ -9,28 +17,111 @@ class NodeType(Enum):
     JOB = auto()
 
 
-class Node:
-    def __init__(self, id_: int, data: Dict, type_: NodeType) -> None:
-        self.parent: Optional[Node] = None
-        self.data: Dict = data
-        self.children: Dict[int, Node] = {}
-        self.id = id_
-        self.type = type_
+@dataclass
+class _Node(ABC):
+    id_: int
+    data: Dict[Any, Any] = field(default_factory=dict)
+    parent: Optional[Union[RootNode, IterNode, RealNode]] = None
+    children: Dict[int, Union[IterNode, RealNode, ForwardModelNode]] = field(
+        default_factory=dict
+    )
 
     def __repr__(self) -> str:
         parent = "no " if self.parent is None else ""
         children = "no " if len(self.children) == 0 else f"{len(self.children)} "
-        return f"Node<{self.type}>@{self.id} with {parent}parent and {children}children"
+        return (
+            f"Node<{type(self)}>@{self.id_} with {parent}parent and {children}children"
+        )
+
+    @abstractmethod
+    def add_child(
+        self,
+        node: Union[IterNode, RealNode, ForwardModelNode],
+        node_id: Optional[int] = None,
+    ) -> None:
+        raise NotImplementedError()
```

Comment:
We can just have `pass` here I think, the decorator `abstractmethod` will take care of exceptions like this.